internal/oci/container.go

```diff
@@ -350,14 +391,105 @@ func (c *Container) exitFilePath() string {
 // IsAlive is a function that checks if a container's init pid exists.
 // It is used to check a container state when we don't want a `$runtime state` call
 func (c *Container) IsAlive() bool {
-	if _, err := findprocess.FindProcess(c.state.Pid); err != nil {
-		logrus.Errorf("container %s not running: %v", c.id, err)
+	_, err := c.pid()
+	if err != nil {
+		logrus.Errorf("checking if pid of %s is running failed: %v", c.id, err)
 		return false
 	}
 
 	return true
 }
 
+// Pid is the external, locking function that returns the container's
+// pid, if it's running, and actually the container process.
+func (c *Container) Pid() (int, error) {
+	c.opLock.Lock()
+	defer c.opLock.Unlock()
+	return c.pid()
+}
+
+// pid is the internal, non-locking version of Pid.
+// It checks that we have an InitPid defined in the state, that pid can be found
+// and it is the same process that was originally started by the runtime.
+func (c *Container) pid() (int, error) {
+	if c.state == nil {
+		return -1, errors.New("state not initialized")
+	}
+	pid := c.state.InitPid
+	if pid <= 0 {
+		return -1, errors.New("pid not initialized")
+	}
+
+	// container has stopped (as pid is initialized but the runc state has overwritten it)
+	if c.state.Pid == 0 {
+		return 0, findprocess.ErrNotFound
+	}
+
+	if _, err := findprocess.FindProcess(pid); err != nil {
+		return -1, err
+	}
+
+	if err := c.verifyPid(pid); err != nil {
+		return -1, err
+	}
+	return pid, nil
+}
+
+// findAndReleasePid attempts to find the container's pid
+// and if it's found, release it. In between these steps,
+// it also verifies the pid running is the same as was originally
+// started by the runtime.
+func (c *Container) findAndReleasePid() (bool, error) {
+	pid := c.state.InitPid
+
+	if err := c.verifyPid(pid); err != nil {
+		if !errors.Is(err, errContainerPidGone) {
+			return false, err
+		}
+		return false, nil
+	}
+
+	process, err := findprocess.FindProcess(pid)
```

Comment:
same here

Reply:
I am not actually sure what house keeping process.Release() does, but it was there before. Without this call, we don't know what process we have to release. Do you think we should drop process.Release() here too?